### PR TITLE
EN-9323: esdt nft api endpoint

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Added a new API endpoint useful for fetching ESDT NFT token data from nodes

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: process/nodeStatusProcessor.go [hunk 1]
```diff
@@ -99,6 +102,30 @@ func (nsp *NodeStatusProcessor) GetNetworkConfigMetrics() (*data.GenericAPIRespo
 	return nil, ErrSendingRequest
 }
 
+// GetAllIssuedESDTs will simply forward all the issued ESDTs from an observer in the metachain
+func (nsp *NodeStatusProcessor) GetAllIssuedESDTs() (*data.GenericAPIResponse, error) {
```
Comment (miiu96) [inline process/nodeStatusProcessor.go:106]: add unit tests for this new method

File: process/nodeStatusProcessor.go [hunk 2]
```diff
@@ -99,6 +102,30 @@ func (nsp *NodeStatusProcessor) GetNetworkConfigMetrics() (*data.GenericAPIRespo
...
+	for _, observer := range observers {
+		var responseAllIssuedESDTs *data.GenericAPIResponse
+
+		_, err := nsp.proc.CallGetRestEndPoint(observer.Address, AllIssuedESDTsPath, &responseAllIssuedESDTs)
+		if err != nil {
+			log.Error("all issued esdts request", "observer", observer.Address, "error", err.Error())
+			continue
+		}
+
+		log.Info("all issued esdts request", "shard id", observer.ShardId, "observer", observer.Address)
```
Comment (miiu96) [inline process/nodeStatusProcessor.go:121]: shard ID

File: process/accountProcessor.go [hunk 3]
```diff
@@ -132,6 +132,38 @@ func (ap *AccountProcessor) GetESDTTokenData(address string, key string) (*data.
...
+				"observer", observer.Address,
+				"http code", respCode)
+			if apiResponse.Error != "" {
+				return nil, errors.New(apiResponse.Error)
+			}
+
+			return &apiResponse, nil
+		}
+
+		log.Error("account get ESDT nft token data error", "observer", observer.Address, "address", address, "error", err.Error())
```
Comment (miiu96) [inline process/accountProcessor.go:161]: is this message you will have word "error" 2 times

File: process/accountProcessor.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func (ap *AccountProcessor) GetESDTTokenData(address string, key string) (*data.
 		apiPath := AddressPath + address + "/esdt/" + key
 		respCode, err := ap.proc.CallGetRestEndPoint(observer.Address, apiPath, &apiResponse)
 		if err == nil || respCode == http.StatusBadRequest || respCode == http.StatusInternalServerError {
-			log.Info("account all ESDT token data error",
+			log.Info("account ESDT token data error",
```
Comment (miiu96) [inline process/accountProcessor.go:116]: remove work "error"

File: process/accountProcessor.go [hunk 5]
```diff
@@ -132,6 +132,38 @@ func (ap *AccountProcessor) GetESDTTokenData(address string, key string) (*data.
...
+		return nil, err
+	}
+
+	for _, observer := range observers {
+		apiResponse := data.GenericAPIResponse{}
+		nonceAsString := fmt.Sprintf("%d", nonce)
+		apiPath := AddressPath + address + "/nft/" + key + "/nonce" + nonceAsString
+		respCode, err := ap.proc.CallGetRestEndPoint(observer.Address, apiPath, &apiResponse)
+		if err == nil || respCode == http.StatusBadRequest || respCode == http.StatusInternalServerError {
+			log.Info("account ESDT NFT token data error",
```
Comment (miiu96) [inline process/accountProcessor.go:148]: remove work "error"

File: api/groups/baseAccountsGroup.go [hunk 6]
```diff
@@ -254,6 +255,53 @@ func (group *accountsGroup) getESDTTokenData(c *gin.Context) {
 	c.JSON(http.StatusOK, esdtTokenResponse)
 }
 
+// getESDTNftTokenData returns the esdt nft data for the given address, esdt token and nonce
+func (group *accountsGroup) getESDTNftTokenData(c *gin.Context) {
+	addr := c.Param("address")
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseAccountsGroup.go:260]: no need. It's just a simple get with no necessary conversions. I will leave it as it is

File: observer/circularQueueNodesProvider.go [hunk 7]
```diff
@@ -38,6 +38,7 @@ func NewCircularQueueNodesProvider(observers []*data.NodeData, configurationFile
 func (cqnp *circularQueueNodesProvider) GetNodesByShardId(shardId uint32) ([]*data.NodeData, error) {
 	cqnp.mutNodes.Lock()
 	defer cqnp.mutNodes.Unlock()
+
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply observer/circularQueueNodesProvider.go:41]: :+1:

File: process/nodeStatusProcessor.go (hunk 1 above)
Comment (bogdan-rosianu) [reply process/nodeStatusProcessor.go:106]: added

File: process/nodeStatusProcessor.go (hunk 2 above)
Comment (bogdan-rosianu) [reply process/nodeStatusProcessor.go:121]: done

File: process/accountProcessor.go (hunk 3 above)
Comment (bogdan-rosianu) [reply process/accountProcessor.go:161]: fixed

File: process/accountProcessor.go (hunk 4 above)
Comment (bogdan-rosianu) [reply process/accountProcessor.go:116]: done

File: process/accountProcessor.go (hunk 5 above)
Comment (bogdan-rosianu) [reply process/accountProcessor.go:148]: done

**Review by miiu96 — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Pull request merged**